// @sap/ux-specification/dist/specification/src/sync/common/ManifestDrivenSchemaProcessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.manifestDrivenSchemaProcessor = void 0;
const rules_1 = require("./rules");
/**
 * Registry for schema processing rules.
 * Stores and resolves rules by name for later execution by the processor.
 */
class ManifestDrivenRuleRegistry {
    constructor() {
        this.rules = new Map();
    }
    /**
     * Registers a new rule by adding it to the collection of rules.
     *
     * @param {NamedRule} rule - The rule to register, including its name and associated functionality.
     * @returns {void}
     */
    register(rule) {
        this.rules.set(rule.name, rule);
    }
    /**
     * Retrieves a named rule from the collection of rules.
     *
     * @param {RuleName} name - The name of the rule to retrieve.
     * @returns {NamedRule | undefined} The corresponding named rule if it exists, or undefined if it does not.
     */
    get(name) {
        return this.rules.get(name);
    }
}
/**
 * Applies schema adaptation rules driven by manifest settings.
 * Given a rule name, page context, and schema, it executes the matching rule(s)
 * and mutates the schema accordingly
 */
class ManifestDrivenSchemaProcessor {
    /**
     * Constructs an instance of the class.
     *
     * @param {ManifestDrivenRuleRegistry} registry - The registry containing the manifest-driven rules.
     * @returns {void}
     */
    constructor(registry) {
        this.registry = registry;
    }
    /**
     * Applies a specific rule from the registry to the given context.
     *
     * @param {RuleName} ruleName - The name of the rule to be applied.
     * @param {RuleContext} context - The context in which the rule will be processed.
     * @returns {void} The method does not return a value.
     */
    apply(ruleName, context) {
        const rule = this.registry.get(ruleName);
        if (!rule) {
            return;
        }
        if (rule.match(context)) {
            rule.process(context);
        }
    }
}
const registry = new ManifestDrivenRuleRegistry();
registry.register(rules_1.LR_CREATION_MODE_FOR_ANALYTICAL_TABLE);
registry.register(rules_1.LR_CREATION_MODE_FOR_NON_TREE_TABLE);
registry.register(rules_1.LR_CREATION_MODE_FOR_NON_CREATION_DIALOG_MODE);
registry.register(rules_1.LR_ANALYTICAL_CONFIGURATION_FOR_NON_ANALYTICAL_TABLE);
exports.manifestDrivenSchemaProcessor = new ManifestDrivenSchemaProcessor(registry);
